vowpalwabbit/learner.cc

```diff
@@ -142,16 +147,23 @@ class single_example_handler
 public:
   single_example_handler(const context_type& context) : _context(context) {}
 
-  void on_example(example* ec)
+  void on_example(std::vector<example*>* ev)
   {
-    if (ec->indices.size() > 1)  // 1+ nonconstant feature. (most common case first)
-      _context.template process<example, learn_ex>(*ec);
-    else if (ec->end_pass)
-      _context.template process<example, end_pass>(*ec);
-    else if (is_save_cmd(ec))
-      _context.template process<example, save>(*ec);
-    else
-      _context.template process<example, learn_ex>(*ec);
+    for (example* ec : *ev)
```

Comment:
nit: technically this is unnecessary since the vector is guaranteed to have exactly 1 element at this point. We can leave it alone for now, but directly indexing the first value might be slightly faster (depends on how well the compiler optimizes this stuff)